crates/compiler/can/src/derive.rs

```diff
@@ -0,0 +1,249 @@
+    let alloc_expr = |it| env.arena.alloc(Loc::at(DERIVED_REGION, it));
+    let hasher = "#hasher";
+
+    let payload = "#payload";
+
+    // \@Opaq payload
+    let opaque_ref = alloc_pat(ast::Pattern::OpaqueRef(at_opaque));
+    let opaque_apply_pattern = ast::Pattern::Apply(
+        opaque_ref,
+        &*env
+            .arena
+            .alloc([Loc::at(DERIVED_REGION, ast::Pattern::Identifier(payload))]),
+    );
+
+    // Hash.hash hasher payload
+    let call_member = alloc_expr(ast::Expr::Apply(
+        alloc_expr(ast::Expr::Var {
+            module_name: "Hash",
+            ident: "hash",
+        }),
+        &*env.arena.alloc([
+            &*alloc_expr(ast::Expr::Var {
+                module_name: "",
+                ident: hasher,
+            }),
+            &*alloc_expr(ast::Expr::Var {
+                module_name: "",
+                ident: payload,
+            }),
+        ]),
+        roc_module::called_via::CalledVia::Space,
+    ));
+
+    // \hasher, @Opaq payload -> Hash.hash hasher payload
+    ast::Expr::Closure(
+        env.arena.alloc([
+            Loc::at(DERIVED_REGION, ast::Pattern::Identifier(hasher)),
+            Loc::at(DERIVED_REGION, opaque_apply_pattern),
+        ]),
+        call_member,
+    )
+}
+
+fn is_eq<'a>(env: &mut Env<'a>, at_opaque: &'a str) -> ast::Expr<'a> {
+    let alloc_pat = |it| env.arena.alloc(Loc::at(DERIVED_REGION, it));
+    let alloc_expr = |it| env.arena.alloc(Loc::at(DERIVED_REGION, it));
+
+    let payload1 = "#payload1";
+    let payload2 = "#payload2";
+
+    let opaque_ref = alloc_pat(ast::Pattern::OpaqueRef(at_opaque));
+    // \@Opaq payload1
+    let opaque1 = ast::Pattern::Apply(
+        opaque_ref,
+        &*env
+            .arena
+            .alloc([Loc::at(DERIVED_REGION, ast::Pattern::Identifier(payload1))]),
+    );
+    // \@Opaq payload2
+    let opaque2 = ast::Pattern::Apply(
+        opaque_ref,
+        &*env
+            .arena
+            .alloc([Loc::at(DERIVED_REGION, ast::Pattern::Identifier(payload2))]),
+    );
+
+    // Bool.isEq payload1 payload2
+    let call_member = alloc_expr(ast::Expr::Apply(
+        alloc_expr(ast::Expr::Var {
+            module_name: "Bool",
+            ident: "isEq",
+        }),
+        &*env.arena.alloc([
+            &*alloc_expr(ast::Expr::Var {
+                module_name: "",
+                ident: payload1,
+            }),
+            &*alloc_expr(ast::Expr::Var {
+                module_name: "",
+                ident: payload2,
+            }),
+        ]),
+        roc_module::called_via::CalledVia::Space,
+    ));
+
+    // \@Opaq payload1, @Opaq payload2 -> Bool.isEq payload1 payload2
+    ast::Expr::Closure(
+        env.arena.alloc([
+            Loc::at(DERIVED_REGION, opaque1),
+            Loc::at(DERIVED_REGION, opaque2),
+        ]),
+        call_member,
+    )
+}
+
+pub const DERIVED_REGION: Region = Region::zero();
```

Comment:
Any particular reason to have this named? (Not a problem, just curious.)

Reply:
Two reasons:
- Harder to make a mistake about the region in the derived impl
- Just in case we decide to change it later, for whatever reason

I think it's a reasonable convention to say region-zero errors shouldn't be reported, but we also have bugs currently where we store some user errors at Region zero, so I think it's good for us to be explicit about it.